runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/GradleDokkaSourceSetBuilder.kt

```diff
@@ -0,0 +1,218 @@
+@file:Suppress("FunctionName", "UnstableApiUsage")
+
+package org.jetbrains.dokka.gradle
+
+import com.android.build.gradle.api.AndroidSourceSet
+import groovy.lang.Closure
+import org.gradle.api.Action
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.provider.ListProperty
+import org.gradle.api.provider.Property
+import org.gradle.api.provider.SetProperty
+import org.gradle.api.tasks.*
+import org.gradle.kotlin.dsl.listProperty
+import org.gradle.kotlin.dsl.setProperty
+import org.gradle.util.ConfigureUtil
+import org.jetbrains.dokka.*
+import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
+import java.io.File
+import java.net.URL
+
+internal fun Task.GradleDokkaSourceSetBuilderFactory(): (name: String) -> GradleDokkaSourceSetBuilder =
+    { name -> GradleDokkaSourceSetBuilder(name, project) }
+
+
+open class GradleDokkaSourceSetBuilder constructor(
+    @Transient @get:Input val name: String,
+    @Transient @get:Internal internal val project: Project
+) : DokkaConfigurationBuilder<DokkaSourceSetImpl> {
+
+    @Internal
+    val sourceSetID: DokkaSourceSetID = DokkaSourceSetID(project, name)
+
+    @Classpath
+    @Optional
+    val classpath: ConfigurableFileCollection = project.files()
+
+    @Input
+    @Optional
+    val moduleDisplayName: Property<String?> = project.objects.safeProperty()
+
+    @Input
+    @Optional
+    val displayName: Property<String?> = project.objects.safeProperty()
+
+    @InputFiles
+    val sourceRoots: ConfigurableFileCollection = project.files()
```

Comment:
Instead of `objects.fileCollection()` would be the better API if you just need an empty collection (If it is available in the Gradle version you want to support).

Even better would be to make the class and all the properties abstract. Like:
```
    @get:InputFiles
    abstract val sourceRoots: ConfigurableFileCollection
```

Then Gradle will inject everything and you have much less boilerplate. And you probably won't have to pass `project` and/or `objects` around.

Reply:
Thanks a lot, I did not know about `objects.fileCollection` and I also did not know about the fact, that Gradle will inject into Abstract classes here as well, which is awesome!! 🤪

Pushed a commit for `objects.fileCollection`. Making this object abstract will require some tests to be adapted. I will wait for another PR to be merged first to tackle this! :)